Mathlib/Order/OrderType.lean

```diff
@@ -0,0 +1,450 @@
+/-
+Copyright (c) _
+-/
+import Mathlib.SetTheory.Cardinal.Basic
+import Mathlib.SetTheory.Ordinal.Basic
+import Mathlib.Order.Category.LinOrd
+import Mathlib.Algebra.Order.Ring.Unbundled.Rat
+import Mathlib.Data.Real.Basic
+import Mathlib.Data.Sum.Order
+import Mathlib.Data.Prod.Lex
+import Mathlib.Logic.Equiv.Prod
+/-!
+# OrderTypes
+
+OrderTypes are defined as equivalences of linear orders under order isomorphism. They are endowed
+with a preorder, where an OrderType is smaller than another one there is an order embedding
+into it.
+
+## Main definitions
+
+* `OrderType`: the type of OrderTypes (in a given universe)
+* `OrderType.type α`: given a type `α` with a linear order, this is the corresponding OrderType,
+* `OrderType.card o`: the cardinality of an OrderType `o`.
+* `OrderType.addMonoid`: the additive monoid of OrderTypes.
+* `OrderType.mul o₁ o₂`: the product of two OrderTypes `o₁` and `o₂`.
+
+A pre order with a bottom element is registered on OrderTypes, where `⊥` is
+`0`, the OrderType corresponding to the empty type.
+
+## Notation
+
+* `ω` is a notation for the order type of `ℕ` with its natural order.
+* `η` is a notation for the order type of `ℚ` with its natural order.
+* `θ` is a notation for the order type of the real numbers on the interval `(0,1)`.
+
+## References
+
+* <https://en.wikipedia.org/wiki/Order_type>
+* Dauben, J. W. Georg Cantor: His Mathematics and Philosophy of the Infinite. Princeton,
+  NJ: Princeton University Press, 1990.
+* Enderton, Herbert B.. Elements of Set Theory. United Kingdom: Academic Press, 1977.
+
+## Tags
+
+order type, order isomorphism, linear order
+-/
+
+noncomputable section
+
+open Function Cardinal Set Equiv Order
+
+universe u v w w'
+
+variable {α : Type u} {β : Type v} {γ : Type w} {δ : Type w'}
+
+def ordIsoOfIsEmpty (α : Type u) (β : Type v) [LinearOrder α] [LinearOrder β]
+    [IsEmpty β] [IsEmpty α] : α ≃o β :=
+  ⟨Equiv.equivOfIsEmpty α β, @fun a ↦ isEmptyElim a⟩
+
+def OrderType.ofUniqueOfIrrefl [LinearOrder α]
+    [LinearOrder β] [Unique α] [Unique β] : α ≃o β :=
+  ⟨Equiv.ofUnique α β, by simp⟩
+
+/-- Equivalence relation on linear orders on arbitrary types in universe `u`, given by order
+isomorphism. -/
+instance OrderType.instSetoid : Setoid LinOrd where
+  r := fun lin_ord₁ lin_ord₂ ↦ Nonempty (lin_ord₁ ≃o lin_ord₂)
+  iseqv := ⟨fun _ ↦ ⟨.refl _⟩, fun ⟨e⟩ ↦ ⟨e.symm⟩, fun ⟨e₁⟩ ⟨e₂⟩ ↦ ⟨e₁.trans e₂⟩⟩
+
+/-- `OrderType.{u}` is the type of linear orders in `Type u`, up to order isomorphism. -/
+@[pp_with_univ]
+def OrderType : Type (u + 1) :=
+  Quotient OrderType.instSetoid
+
+namespace OrderType
+
+def toType (o : OrderType) : Type u :=
+  o.out.carrier
+
+instance (o : OrderType) : LinearOrder o.toType :=
+  o.out.str
+
+/-! ### Basic properties of the order type -/
+
+/-- The order type of a linear order on α. -/
+def type (α : Type u) [LinearOrder α] : OrderType :=
+  ⟦⟨α⟩⟧
+
+instance zero : Zero OrderType where
```

Comment:
About two years ago, there was some discussion which ultimately culminated in changing the definition of the `Zero Cardinal` instance from `#PEmpty = 0` to #(Fin 0) = 0`. The idea here being that in pen and paper mathematics, one defines the cardinal `n` as the set of ordinals/naturals below `n`.

I don't really agree with this reasoning, and I think it's fine to just define the zero/one cardinals/ordinals using `PEmpty` and `PUnit`, but I figured you should be aware of this.